File: src/openrtm_aist/src/lib/coil/common/TimeMeasure.cpp
// -*- C++ -*-
/*!
 * @file TimeMeasure.cpp
 * @brief Periodic time measurement class
 * @date $Date$
 * @author Noriaki Ando <n-ando@aist.go.jp>
 *
 * Copyright (C) 2009
 *     Noriaki Ando
 *     Task-intelligence Research Group,
 *     Intelligent Systems Research Institute,
 *     National Institute of
 *         Advanced Industrial Science and Technology (AIST), Japan
 *     All rights reserved.
 *
 * $Id$
 *
 */

#include <coil/Time.h>
#include <math.h>
#include <coil/TimeMeasure.h>

//#define RDTSC(X) __asm__ __volatile__ ("rdtsc" : "=A" (X))
//#define NSEC_PER_SEC 1000000000

#ifndef ULLONG_MAX
#define ULLONG_MAX 0xffffffffffffffffULL
#endif 

namespace coil
{

  /*!
   * @if jp
   * @brief コンストラクタ
   * @else
   * @brief Constructor
   * @endif
   */
  TimeMeasure::TimeMeasure(int buflen)
    : m_begin(0.0), m_interval(0.0),
      m_count(0), m_countMax(buflen + 1),
      m_recurred(false)
  {
    m_record.reserve(m_countMax);
    for (unsigned long i(0); i < m_countMax; ++i)
      {
        m_record.push_back(TimeValue(0, 0));
      }
  }

  /*!
   * @if jp
   * @brief 時間統計の計測を開始する
   * @else
   * @brief Begin time measurement for time statistics
   * @endif
   */
  void TimeMeasure::tick()
  {
    m_begin = gettimeofday(); // [TimeValue]
  }

  /*!
   * @if jp
   * @brief 時間統計の計測を終了する
   * @else
   * @brief Finish time measurement for time statistics
   * @endif
   */
  void TimeMeasure::tack()
  {
    if (m_begin.sec() == 0) { return; }

    m_interval = gettimeofday() - m_begin;
    m_record.at(m_count) = m_interval;
    ++m_count;
    if (m_count == m_countMax)
      {
        m_count = 0;
        m_recurred = true;
      }
  }

  /*!
   * @if jp
   * @brief 経過時間を取得する
   * @else
   * @brief Get a interval time
   * @endif
   */
  coil::TimeValue& TimeMeasure::interval()
  {
    return m_interval;
  }

  /*!
   * @if jp
   * @brief 統計関連データの初期化
   * @else
   * @brief Initialize for statistics related data
   * @endif
   */
  void TimeMeasure::reset()
  {
    m_count = 0;
    m_recurred = false;
    m_begin = 0.0;
  }

  /*!
   * @if jp
   * @brief 時間統計バッファサイズを取得する
   * @else
   * @brief Get number of time measurement buffer
   * @endif
   */
  unsigned long int TimeMeasure::count() const
  {
    return m_recurred ? m_record.size() : m_count;
  }

  /*!
   * @if jp
   * @brief 統計データの総計を取得する
   * @else
   * @brief Get total statistics
   * @endif
   */
  bool TimeMeasure::getStatistics(double &max_interval,
                                  double &min_interval,
                                  double &mean_interval,
                                  double &stddev)
  {
    max_interval = (double)0;
    min_interval = (double)ULLONG_MAX;

    double sum = 0;
    double sq_sum = 0;
    unsigned long int len(count());

    if (len == 0) return false;

    for (unsigned long int i(0); i < len; ++i)
      {
        double trecord(m_record[i]);
        sum += trecord;
        sq_sum += trecord * trecord;
        
        if (trecord > max_interval) max_interval = trecord;
        if (trecord < min_interval) min_interval = trecord;
      }
    
    mean_interval = sum / len;
    stddev = sqrt(sq_sum / len - (mean_interval * mean_interval));

    return true;
  }

  /*!
   * @if jp
   * @brief 統計結果を取得する
   * @else
   * @brief Get statistics result
   * @endif
   */
  TimeMeasure::Statistics TimeMeasure::getStatistics()
  {
    Statistics s;
    getStatistics(s.max_interval, s.min_interval,
                  s.mean_interval, s.std_deviation);
    return s;
  }

}; // namespace coil

File: src/openrtm_aist_python/OpenRTM_aist/TimeMeasure.py
#!/usr/bin/env python
# -*- coding: euc-jp -*-

##
# @file TimeMeasure.py
# @brief Periodic time measurement class
# @date $Date$
# @author Noriaki Ando <n-ando@aist.go.jp> and Shinji Kurihara
#
# Copyright (C) 2009
#     Noriaki Ando
#     Task-intelligence Research Group,
#     Intelligent Systems Research Institute,
#     National Institute of
#         Advanced Industrial Science and Technology (AIST), Japan
#     All rights reserved.
#
# $Id$
#
#

import time
import math

import OpenRTM_aist

ULLONG_MAX = 0xffffffffffffffff

##
# @if jp
# @brief 時間単位変換用定数
# @else
# @endif
usec_per_sec = 1000000

##
# @if jp
# @class Time
# @brief 時間管理用クラス
# 
# 指定した時間値を保持するためのクラス。
# 
# @since 0.4.1
# 
# @else
# 
# @endif
class Time:
  """
  """

  ##
  # @if jp
  # @brief コンストラクタ
  #
  # コンストラクタ。
  #
  # @param self
  #
  # @else
  # @brief Constructor.
  #
  # Constructor.
  #
  # @param self
  #
  # @endif
  def __init__(self):
    global usec_per_sec
    tm = time.time()
    tm_f       = float(tm - long(tm))     # 小数部の取り出し
    self.sec   = long(tm - tm_f)   # 整数部の取り出し
    self.usec  = long(float(tm_f) * float(usec_per_sec)) # sec -> usec (micro second)
    self._timevalue = OpenRTM_aist.TimeValue(self.sec,self.usec)
    return

  ##
  # @if jp
  #
  # @brief 時間減算
  # 
  # 設定された時間から引数で与えられた時間を減算する。
  #
  # @param self
  # @param tm 減算時間
  # 
  # @return 減算結果
  # 
  # @else
  #
  # @endif
  def __sub__(self, tm):
    global usec_per_sec

    res = Time()
    
    if self.sec >= tm.sec:
      if self.usec >= tm.usec:
        res.sec  = self.sec  - tm.sec
        res.usec = self.usec - tm.usec
      else:
        res.sec  = self.sec  - tm.sec - 1
        res.usec = (self.usec + usec_per_sec) - tm.usec
    else:
      if tm.usec >= self.usec:
        res.sec  = -(tm.sec  - self.sec)
        res.usec = -(tm.usec - self.usec)
      else:
        res.sec  = -(tm.sec - self.sec - 1)
        res.usec = -(tm.usec + usec_per_sec) + self.usec
    return res

    
  def gettimeofday(self):
    global usec_per_sec
    tm = time.time()
    tm_f       = float(tm - long(tm))     # 小数部の取り出し
    self.sec   = long(float(tm) - float(tm_f))   # 整数部の取り出し
    self.usec  = long(float(tm_f) * float(usec_per_sec)) # sec -> usec (micro second)
    return OpenRTM_aist.TimeValue(self.sec, self.usec)


  def getTime(self):
    return OpenRTM_aist.TimeValue(self.sec, self.usec)


##
# TimeMeasure object
#
# This object is used for getting statistics of code execution time. 
# Using get_stat you can get maximum, minimum, mean and standard
# deviation time for code execution.
#
class TimeMeasure:
  """
  """

  ##
  # @brief Time statictics object for profiling.
  # 
  # Constructor
  #
  def __init__(self, buflen=100):
    self._countMax = buflen + 1
    self._record = [OpenRTM_aist.TimeValue(0, 0) for i in range(self._countMax)]
    self._begin  = Time().gettimeofday()
    self._end  = Time().gettimeofday()
    self._count = 0
    self._recurred = False
    self._interval = OpenRTM_aist.TimeValue(0.0)
    return

  ##
  # @brief Begin time measurement for time statistics.
  #
  # Begin time measurement for time statistics
  #
  def tick(self):
    self._begin = Time().gettimeofday()
    return

  ##
  # @brief Finish time measurement for time statistics.
  #
  # End of time measurement for time statistics
  #
  def tack(self):
    if self._begin.sec() == 0:
      return

    self._interval = Time().gettimeofday() - self._begin
    self._end = Time().gettimeofday()
    self._record[self._count] = self._interval
    self._count += 1
    if self._count == self._countMax:
      self._count = 0
      self._recurred = True
    return

  def interval(self):
    return self._interval


  def reset(self):
    self._count = 0
    self._recurred = False
    self._begin = OpenRTM_aist.TimeValue(0.0)
    return
    
  ##
  # Get number of time measurement buffer
  #
  # @brief Get number of time measurement buffer.
  #
  #
  def count(self):
    if self._recurred:
      return len(self._record)
    else:
      return self._count
    return
    
  ##
  # @brief Get total statistics.
  # Get total statistics
  # max_interval, min_interval, mean_interval [ns]
  #
  def getStatistics(self, max_interval=None, min_interval=None,
                    mean_interval=None, stddev=None):
    global ULLONG_MAX

    if not max_interval and not min_interval and not mean_interval and not stddev:
      max_i  = [0.0]
      min_i  = [0.0]
      mean_i = [0.0]
      stdd   = [0.0]
      return self.getStatistics(max_i, min_i, mean_i, stdd)

    max_interval[0] = 0.0
    min_interval[0] = ULLONG_MAX

    sum_ = 0.0
    sq_sum_ = 0.0
    len_ = self.count()
        
    if len_ == 0:
      return False

    for i in range(len_):
      trecord_ = self._record[i].toDouble()
      sum_ += trecord_
      sq_sum_ += (trecord_ * trecord_)
      
      if trecord_ > max_interval[0]:
        max_interval[0] = trecord_
            
      if trecord_ < min_interval[0]:
        min_interval[0] = trecord_
            
                
    mean_interval[0] = sum_ / len_
    stddev[0] = math.sqrt(sq_sum_ / len_ - (mean_interval[0] * mean_interval[0]))

    return True
        

  class Statistics:
    def __init__(self, max=None, min=None, mean=None, stdd=None):
      if not max and not min and not mean and not stdd:
        self._max_interval  = 0.0
        self._min_interval  = 0.0
        self._mean_interval = 0.0
        self._std_deviation = 0.0
        return

      self._max_interval  = max
      self._min_interval  = min
      self._mean_interval = mean
      self._std_deviation = stdd
      return
